In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Use the kagglehub client library to attach Kaggle resources like competitions, datasets, and models to your session
# Learn more about kagglehub: https://github.com/Kaggle/kagglehub/blob/main/README.md

import kagglehub
# kagglehub.dataset_download('<owner>/<dataset-slug>')

/kaggle/input/datasets/arshid/iris-flower-dataset/IRIS.csv


# 1. Load and inspect data

In [2]:
df = pd.read_csv('/kaggle/input/datasets/arshid/iris-flower-dataset/IRIS.csv')
df.head()

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [4]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
sepal_length,150.0,5.843333,0.828066,4.3,5.1,5.80,6.4,7.9
sepal_width,150.0,3.054000,0.433594,2.0,2.8,3.00,3.3,4.4
petal_length,150.0,3.758667,1.764420,1.0,1.6,4.35,5.1,6.9
petal_width,150.0,1.198667,0.763161,0.1,0.3,1.30,1.8,2.5


In [5]:
df.isna().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

No missing/null values

In [7]:
df['species'].value_counts()

species
Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
Name: count, dtype: int64

In [8]:
df.dtypes

sepal_length    float64
sepal_width     float64
petal_length    float64
petal_width     float64
species          object
dtype: object

1. Features are in numeric no encoding is required.
2. No missing/null values.
3. We need scaling.
4. All the features are required here.
5. No feature engineering is required in this dataset.
6. So we can directly start with train-test-split, feature scaling and model training.

In [11]:
print(df['species'].nunique())
print(df['species'].unique())

3
['Iris-setosa' 'Iris-versicolor' 'Iris-virginica']


In [12]:
df['species'].value_counts()

species
Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
Name: count, dtype: int64

# 2. Train-test split

In [13]:
from sklearn.model_selection import train_test_split

X = df.drop(columns=['species'])
y = df['species']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42,
    shuffle=True,
    stratify=y
)

print(f"Training features shape: {X_train.shape}")
print(f"Testing features shape: {X_test.shape}")

Training features shape: (120, 4)
Testing features shape: (30, 4)


# 3. Feature Scaling

In [14]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# 4. Model Training

In [17]:
from sklearn.linear_model import LogisticRegression
from sklearn.multiclass import OneVsRestClassifier

# Version A - default (multinomial)
model_default = LogisticRegression()
model_default.fit(X_train_scaled, y_train)
y_pred_default = model_default.predict(X_test_scaled)
print(y_pred_default)

# Version B - explicit OvR
model_ovr = OneVsRestClassifier(LogisticRegression())
model_ovr.fit(X_train_scaled, y_train)
y_pred_ovr = model_ovr.predict(X_test_scaled)
print(y_pred_ovr)

['Iris-setosa' 'Iris-virginica' 'Iris-versicolor' 'Iris-versicolor'
 'Iris-setosa' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-virginica' 'Iris-virginica'
 'Iris-virginica' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-versicolor' 'Iris-versicolor' 'Iris-virginica'
 'Iris-setosa' 'Iris-virginica' 'Iris-versicolor' 'Iris-versicolor'
 'Iris-virginica' 'Iris-virginica' 'Iris-versicolor' 'Iris-setosa'
 'Iris-virginica' 'Iris-setosa']
['Iris-setosa' 'Iris-virginica' 'Iris-versicolor' 'Iris-versicolor'
 'Iris-setosa' 'Iris-virginica' 'Iris-setosa' 'Iris-setosa'
 'Iris-virginica' 'Iris-versicolor' 'Iris-virginica' 'Iris-virginica'
 'Iris-virginica' 'Iris-versicolor' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-versicolor' 'Iris-versicolor' 'Iris-virginica'
 'Iris-setosa' 'Iris-virginica' 'Iris-versicolor' 'Iris-versicolor'
 'Iris-virginica' 'Iris-virginica' 'Iris-versicolor' 'Iris-setosa'
 'Iris-virginica' 'Iris-setosa']


# 5. Evaluation

In [18]:
from sklearn.metrics import accuracy_score
print(f'Default: {accuracy_score(y_test, y_pred_default)}')
print(f'Explicit OvR: {accuracy_score(y_test, y_pred_ovr)}')

Default: 0.9333333333333333
Explicit OvR: 0.9


In [19]:
from sklearn.metrics import confusion_matrix

print(f'Default: {confusion_matrix(y_test, y_pred_default)}')
print(f'Explicit OvR: {confusion_matrix(y_test, y_pred_ovr)}')

Default: [[10  0  0]
 [ 0  9  1]
 [ 0  1  9]]
Explicit OvR: [[10  0  0]
 [ 0  8  2]
 [ 0  1  9]]


In [22]:
from sklearn.metrics import precision_score, recall_score, f1_score, classification_report

precision_default = precision_score(y_test, y_pred_default, average='macro')
recall_default = recall_score(y_test, y_pred_default, average='macro')
f1_default = f1_score(y_test, y_pred_default, average='macro')

precision_ovr = precision_score(y_test, y_pred_ovr, average='macro')
recall_ovr = recall_score(y_test, y_pred_ovr, average='macro')
f1_ovr = f1_score(y_test, y_pred_ovr, average='macro')

print(f"Precision Default: {precision_default:.3f}, Precision Ovr: {precision_ovr:.3f}")
print(f"Recall Default: {recall_default:.3f}, Recall Ovr: {recall_ovr:.3f}")
print(f"F1-score Default: {f1_default:.3f}, F1-score Ovr: {f1_ovr:.3f}")

print(classification_report(y_test, y_pred_default))
print(classification_report(y_test, y_pred_ovr))

Precision Default: 0.933, Precision Ovr: 0.902
Recall Default: 0.933, Recall Ovr: 0.900
F1-score Default: 0.933, F1-score Ovr: 0.900
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        10
Iris-versicolor       0.90      0.90      0.90        10
 Iris-virginica       0.90      0.90      0.90        10

       accuracy                           0.93        30
      macro avg       0.93      0.93      0.93        30
   weighted avg       0.93      0.93      0.93        30

                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        10
Iris-versicolor       0.89      0.80      0.84        10
 Iris-virginica       0.82      0.90      0.86        10

       accuracy                           0.90        30
      macro avg       0.90      0.90      0.90        30
   weighted avg       0.90      0.90      0.90        30



## Multi-class Classification — Iris Dataset (OvR vs Multinomial)

### Key Learnings
- Check the target column's unique class count: 2 classes → binary classification is enough; 3+ classes → need a multi-class strategy (OvR, OvO, or Multinomial/Softmax)
- sklearn's default `LogisticRegression()` already uses multinomial (softmax) internally — to force explicit OvR, use the `OneVsRestClassifier(LogisticRegression())` wrapper (the old `multi_class='ovr'` parameter is deprecated as of sklearn 1.5)
- Class balance/imbalance is a **separate question** from binary vs multi-class — it only determines how much extra care is needed (stratification, per-class metrics), not whether classification is possible
- Multi-class metrics (`precision_score`, `recall_score`, `f1_score`) require an `average` parameter (`'macro'`, `'weighted'`, `'micro'`, or `None`) — the binary default won't work with 3+ classes
- `'macro'` average treats all classes equally; `'weighted'` accounts for class sample sizes — for a balanced dataset like Iris, both give similar results

### Results Comparison

| Metric | Multinomial (default) | OvR |
|---|---|---|
| Accuracy | 93.3% | 90.0% |
| Macro Precision | 0.933 | 0.902 |
| Macro Recall | 0.933 | 0.900 |
| Macro F1-score | 0.933 | 0.900 |

**Per-class breakdown:**
- `Iris-setosa`: Perfect score (1.00 precision/recall) in both models — this species is linearly well-separated from the other two
- `Iris-versicolor` / `Iris-virginica`: Both models struggle more here due to feature overlap between these two species. OvR performed noticeably worse on `versicolor` recall (0.80 vs 0.90), likely because it evaluates each class independently ("versicolor vs rest") rather than jointly optimizing across all classes like multinomial does.

### Takeaway
Multinomial (softmax) is generally the better default for Logistic Regression on multi-class problems, since it optimizes a single unified decision boundary across all classes simultaneously. OvR remains most useful for algorithms that don't natively support multinomial outputs (e.g., SVM), or when per-class model interpretability is specifically needed.

### End-to-End Practice Summary (Both Sessions)
1. **Titanic dataset** (binary, imbalanced) → missing value imputation, encoding strategy (one-hot vs as-is), feature engineering (FamilySize, Title extraction), data leakage prevention, stratified splitting, accuracy improved from 82.1% → 84.9% after feature engineering
2. **Iris dataset** (multi-class, balanced) → applied the same core pipeline (split → scale → train → evaluate), compared Multinomial vs OvR strategies, learned when and why to choose one over the other